src/vaults/trading-apys/implementations/univ3-seth2.js

```diff
@@ -0,0 +1,52 @@
+const { web3Socket } = require('../../../lib/web3')
+const univ3ControllerContract = require('../../../lib/web3/contracts/uniswap-v3-controller/contract.json')
+
+// fromBlock = 12429930: It was the earliest block when Uniswap V3 vaults were deployed
+const getTradingApy = async (vaultAddress, fromBlock = 12429930, toBlock = 'latest') => {
+  let latestHarvestsToAverageOverForVault,
+    latestHarvestsToAverageOver = 3,
+    dailyAPRTotal = 0
+
+  const instance = new web3Socket.eth.Contract(
+    univ3ControllerContract.abi,
+    univ3ControllerContract.address.mainnet,
+  )
+  const vaultEvents = (
+    await instance.getPastEvents('SharePriceChangeTrading', {
+      filter: {
+        vault: vaultAddress,
```

Comment:
This event doesn't have the field `vault`. Is there a generic way in web3 to check which address the event is emitted upon?

Reply:
So is the event emitted on the vault or on the controller. Because in the current setup it is reading the events on the controller. We could change that to read on the vault contracts, but that would need some new logic. It would need the vault contract ABI and some input for the vault address.

Reply:
The vault's code `emit`s this event, but the transaction is on the controller, like all other doHardWorks.
My point is that `SharePriceChangeLog` has `vault` as a field explicitly:
```
  event SharePriceChangeLog(
    address indexed vault,
    address indexed strategy,
    uint256 oldSharePrice,
    uint256 newSharePrice,
    uint256 timestamp
  );
```

While `SharePriceChangeTrading` and `SharePriceChangeLiquidation` do not have such a field.
So, is there a way to figure out the vault address in the current setup? If yes, we'll need to change the current code (since there is no `vault` field in the event). If no, we will need to upgrade the vault and add such a field.

Reply:
Hey, yeah so a few things need to be changed, but it will not be a problem. I will have an update probably tomorrow. The difference is that is the original case which I copied from, the event is emitted by the controller. So by just changing the lookup address for the event to the vault address, we will automatically be filtering the events per vault.

Reply:
ok, let's try it out!